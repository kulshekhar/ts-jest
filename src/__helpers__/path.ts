import { ensureDirSync } from 'fs-extra'
import { tmpdir } from 'os'
import { join, resolve } from 'path'

export const ROOT = resolve(__dirname, '..', '..')

export function tempDir(ns: string): string {
  const dir = join(tmpdir(), 'ts-jest-unit-tests', ns, parseInt((Math.random() * 1e17) as any, 10).toString(36))
  ensureDirSync(dir)

  return dir
}
