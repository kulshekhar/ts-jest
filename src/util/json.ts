/* eslint-disable no-redeclare */
import stableStringify = require('fast-json-stable-stringify')

const UNDEFINED = 'undefined'

/**
 * @internal
 */
export function stringify(input: any): string {
  return input === undefined ? UNDEFINED : stableStringify(input)
}

/**
 * @internal
 */
export function parse(input: string): any {
  return input === UNDEFINED ? undefined : JSON.parse(input)
}

interface NormalizeOptions {
  parse?: (input: string) => any
}
/**
 * @internal
 */
export function normalize(input: string, { parse: parser = parse }: NormalizeOptions = {}): string {
  let result: string | undefined
  if (normalize.cache.has(input)) {
    result = normalize.cache.get(input)
  } else {
    const data = parser(input)
    result = stringify(data)
    if (result === input) result = undefined
    normalize.cache.set(input, result)
  }
  return result === undefined ? input : result
}

/**
 * @internal
 */
// eslint-disable-next-line @typescript-eslint/no-namespace
export namespace normalize {
  export const cache = new Map<string, string | undefined>()
}
