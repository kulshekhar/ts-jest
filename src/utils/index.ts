export { mocked } from './testing'
export { createJestPreset } from '../presets/create-jest-preset'
export { pathsToModuleNameMapper } from '../config/paths-to-module-name-mapper'
