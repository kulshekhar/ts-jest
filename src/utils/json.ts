/* eslint-disable no-redeclare */
import stableStringify = require('fast-json-stable-stringify')

const UNDEFINED = 'undefined'

export function stringify(input: unknown): string {
  return input === undefined ? UNDEFINED : stableStringify(input)
}

export function parse(input: string): any {
  return input === UNDEFINED ? undefined : JSON.parse(input)
}

interface NormalizeOptions {
  parse?: (input: string) => any
}
/**
 * @internal
 */
export function normalize(input: string, { parse: parser = parse }: NormalizeOptions = {}): string {
  let result: string | undefined
  if (normalize.cache.has(input)) {
    result = normalize.cache.get(input)
  } else {
    const data = parser(input)
    result = stringify(data)
    if (result === input) result = undefined
    normalize.cache.set(input, result)
  }

  return result === undefined ? input : result
}

/**
 * @internal
 */
// eslint-disable-next-line @typescript-eslint/no-namespace
export namespace normalize {
  export const cache = new Map<string, string | undefined>()
}
