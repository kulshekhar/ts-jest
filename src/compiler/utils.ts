import * as path from 'path';
import * as ts from 'typescript';
import * as fs from 'fs';

// from here: https://github.com/s-panferov/awesome-typescript-loader/blob/master/src/helpers.ts

export interface OutputFile {
  text: string;
  sourceMap?: string;
  declaration?: ts.OutputFile;
}

function statSyncNoException(path: string) {
  try {
    return fs.statSync(path);
  } catch (e) {
    return undefined;
  }
}

function withoutExt(fileName: string): string {
  return path.basename(fileName).split('.')[0];
}

let caseInsensitiveFs: boolean | undefined;

export function isCaseInsensitive() {
  if (typeof caseInsensitiveFs !== 'undefined') {
    return caseInsensitiveFs;
  }

  const lowerCaseStat = statSyncNoException(process.execPath.toLowerCase());
  const upperCaseStat = statSyncNoException(process.execPath.toUpperCase());

  if (lowerCaseStat && upperCaseStat) {
    caseInsensitiveFs =
      lowerCaseStat.dev === upperCaseStat.dev &&
      lowerCaseStat.ino === upperCaseStat.ino;
  } else {
    caseInsensitiveFs = false;
  }

  return caseInsensitiveFs;
}

function compareFileName(first: string, second: string) {
  if (isCaseInsensitive()) {
    return first.toLowerCase() === second.toLowerCase();
  } else {
    return first === second;
  }
}

function isFileEmit(
  fileName: string,
  outputFileName: string,
  sourceFileName: string,
) {
  return (
    compareFileName(sourceFileName, fileName) &&
    // typescript now emits .jsx files for .tsx files.
    (outputFileName.substr(-3).toLowerCase() === '.js' ||
      outputFileName.substr(-4).toLowerCase() === '.jsx')
  );
}

function isSourceMapEmit(
  fileName: string,
  outputFileName: string,
  sourceFileName: string,
) {
  return (
    compareFileName(sourceFileName, fileName) &&
    // typescript now emits .jsx files for .tsx files.
    (outputFileName.substr(-7).toLowerCase() === '.js.map' ||
      outputFileName.substr(-8).toLowerCase() === '.jsx.map')
  );
}

function isDeclarationEmit(
  fileName: string,
  outputFileName: string,
  sourceFileName: string,
) {
  return (
    compareFileName(sourceFileName, fileName) &&
    outputFileName.substr(-5).toLowerCase() === '.d.ts'
  );
}

export function findResultFor(
  fileName: string,
  output: ts.EmitOutput,
): OutputFile {
  let text: string | undefined;
  let sourceMap: string | undefined;
  let declaration: ts.OutputFile | undefined;
  fileName = withoutExt(fileName);

  for (let i = 0; i < output.outputFiles.length; i++) {
    let o = output.outputFiles[i];
    let outputFileName = o.name;
    let sourceFileName = withoutExt(o.name);
    if (isFileEmit(fileName, outputFileName, sourceFileName)) {
      text = o.text;
    }
    if (isSourceMapEmit(fileName, outputFileName, sourceFileName)) {
      sourceMap = o.text;
    }
    if (isDeclarationEmit(fileName, outputFileName, sourceFileName)) {
      declaration = o;
    }
  }
  if (text === undefined) {
    // console.log(text)
    throw new Error('text is undefined');
  }
  return {
    text,
    sourceMap,
    declaration,
  };
}
