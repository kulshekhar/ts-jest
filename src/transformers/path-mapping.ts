/**
 * This transformer is heavily inspired from:
 * - https://github.com/LeDDGroup/typescript-transform-paths
 * - https://github.com/dropbox/ts-transform-import-path-rewrite
 * Thank you: @longlho, @LeddGroup for all the great works
 */
import { LogContexts, LogLevels } from 'bs-logger'
import { basename, dirname, isAbsolute, join, normalize, relative } from 'path'
import type * as _ts from 'typescript'

import type { ConfigSet } from '../config/config-set'

/**
 * @internal
 */
export const name = 'path-mapping'
// increment this each time the code is modified
/**
 * @internal
 */
export const version = 1

const isBaseDir = (base: string, dir: string) => !relative(base, dir)?.startsWith('.')

/**
 * The factory of import path alias transformer factory.
 */
export function factory(cs: ConfigSet): (ctx: _ts.TransformationContext) => _ts.Transformer<_ts.SourceFile> {
  const logger = cs.logger.child({ namespace: 'ts-path-mapping' })
  const ts = cs.compilerModule
  const compilerOptions = cs.parsedTsConfig.options
  const rootDirs = compilerOptions.rootDirs?.filter(isAbsolute)

  const isDynamicImport = (node: _ts.Node): node is _ts.CallExpression =>
    ts.isCallExpression(node) && node.expression.kind === ts.SyntaxKind.ImportKeyword
  const isRequire = (node: _ts.Node): node is _ts.CallExpression =>
    ts.isCallExpression(node) &&
    ts.isIdentifier(node.expression) &&
    node.expression.text === 'require' &&
    ts.isStringLiteral(node.arguments[0]) &&
    node.arguments.length === 1

  const createVisitor = (ctx: _ts.TransformationContext, sf: _ts.SourceFile) => {
    const { fileName } = sf
    const fileDir = normalize(dirname(fileName))
    const rewritePath = (importPath: string): string => {
      let p = importPath
      const { resolvedModule } = ts.resolveModuleName(importPath, fileName, compilerOptions, ts.sys)
      if (resolvedModule) {
        const { resolvedFileName } = resolvedModule
        let filePath = fileDir
        let modulePath = dirname(resolvedFileName)
        /* Handle rootDirs mapping */
        if (rootDirs) {
          let fileRootDir = ''
          let moduleRootDir = ''
          for (const rootDir of rootDirs) {
            if (isBaseDir(rootDir, resolvedFileName) && rootDir.length > moduleRootDir.length) moduleRootDir = rootDir
            if (isBaseDir(rootDir, fileName) && rootDir.length > fileRootDir.length) fileRootDir = rootDir
          }
          /* Remove base dirs to make relative to root */
          if (fileRootDir && moduleRootDir) {
            filePath = relative(fileRootDir, filePath)
            modulePath = relative(moduleRootDir, modulePath)
          }
        }
        p = normalize(join(relative(filePath, modulePath), basename(resolvedFileName)))
        p = p.startsWith('.') ? p : `./${p}`
      }

      return p
    }
    const visitor: _ts.Visitor = (node) => {
      let rewrittenPath: string | undefined
      const newNode = ts.getMutableClone(node)
      /**
       * e.g.
       * - import('@utils/json')
       * - const { stringify } = require('@utils/json')
       */
      if (isDynamicImport(node) || isRequire(node)) {
        rewrittenPath = rewritePath((node.arguments[0] as _ts.StringLiteral).text)

        return {
          ...newNode,
          arguments: ts.createNodeArray([ts.createStringLiteral(rewrittenPath)]),
        }
      }
      // legacy import, e.g. import foo = require('@utils/json')
      if (ts.isExternalModuleReference(node) && ts.isStringLiteral(node.expression)) {
        rewrittenPath = rewritePath(node.expression.text)

        return ts.updateExternalModuleReference(newNode as _ts.ExternalModuleReference, ts.createLiteral(rewrittenPath))
      }
      /**
       * e.g.
       * - import { parse } from '@utils/json'
       * - import * as json from '@utils/json'
       */
      if (ts.isImportDeclaration(node) && ts.isStringLiteral(node.moduleSpecifier)) {
        rewrittenPath = rewritePath(node.moduleSpecifier.text)

        return {
          ...newNode,
          moduleSpecifier: ts.createLiteral(rewrittenPath),
        }
      }
      // e.g. export * as jsonUtils from '@utils/json'
      if (ts.isExportDeclaration(node) && node.moduleSpecifier && ts.isStringLiteral(node.moduleSpecifier)) {
        rewrittenPath = rewritePath(node.moduleSpecifier.text)

        return {
          ...newNode,
          moduleSpecifier: ts.createLiteral(rewrittenPath),
        }
      }
      // 3.8 import type, e.g. import type { Foo } from '@utils/json'
      if (
        ts.isImportTypeNode(node) &&
        ts.isLiteralTypeNode(node.argument) &&
        ts.isStringLiteral(node.argument.literal)
      ) {
        // `.text` instead of `getText` bc this node doesn't map to sf (it's generated d.ts)
        rewrittenPath = rewritePath(node.argument.literal.text)

        return {
          ...newNode,
          argument: ts.createLiteralTypeNode(ts.createStringLiteral(rewrittenPath)),
        }
      }

      return ts.visitEachChild(node, visitor, ctx)
    }

    return visitor
  }

  // returns the transformer factory
  return (ctx: _ts.TransformationContext): _ts.Transformer<_ts.SourceFile> =>
    logger.wrap(
      { [LogContexts.logLevel]: LogLevels.debug, call: null },
      'visitSourceFileNode(): path mapping',
      (sf: _ts.SourceFile) => ts.visitNode(sf, createVisitor(ctx, sf)),
    )
}
