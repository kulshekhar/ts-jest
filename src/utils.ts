import * as crypto from 'crypto';
import * as fs from 'fs';
import * as fsExtra from 'fs-extra';
import * as path from 'path';
import * as tsc from 'typescript';
import { ConfigGlobals, JestConfig, TsJestConfig } from './jest-types';
import { logOnce } from './logger';
import * as _ from 'lodash';

export function getTSJestConfig(globals: ConfigGlobals): TsJestConfig {
  return globals && globals['ts-jest'] ? globals['ts-jest'] : {};
}

function formatTsDiagnostics(errors: tsc.Diagnostic[]): string {
  const defaultFormatHost: tsc.FormatDiagnosticsHost = {
    getCurrentDirectory: () => tsc.sys.getCurrentDirectory(),
    getCanonicalFileName: fileName => fileName,
    getNewLine: () => tsc.sys.newLine,
  };

  return tsc.formatDiagnostics(errors, defaultFormatHost);
}

function readCompilerOptions(
  configPath: string,
  rootDir: string,
): tsc.CompilerOptions {
  configPath = path.resolve(rootDir, configPath);
  const { config, error } = tsc.readConfigFile(configPath, tsc.sys.readFile);
  if (error) {
    throw error;
  }

  const { errors, options } = tsc.parseJsonConfigFileContent(
    config,
    tsc.sys,
    path.resolve(rootDir),
  );

  if (errors.length > 0) {
    const formattedErrors = formatTsDiagnostics(errors);
    throw new Error(
      `Some errors occurred while attempting to read from ${configPath}: ${formattedErrors}`,
    );
  }

  return options;
}

function getStartDir(): string {
  // This is needed because of the way our tests are structured.
  // If this is being executed as a library (under node_modules)
  // we want to start with the project directory that's three
  // levels above.
  // If this is being executed from the test suite, we want to start
  // in the directory of the test

  const grandparent = path.resolve(__dirname, '..', '..');
  if (grandparent.endsWith(`${path.sep}node_modules`)) {
    return process.cwd();
  }

  return '.';
}

function getPathToClosestTSConfig(
  startDir?: string,
  previousDir?: string,
): string {
  // Starting with the startDir directory and moving on to the
  // parent directory recursively (going no further than the root directory)
  // find and return the path to the first encountered tsconfig.json file

  if (!startDir) {
    return getPathToClosestTSConfig(getStartDir());
  }

  const tsConfigPath = path.join(startDir, 'tsconfig.json');

  const startDirPath = path.resolve(startDir);
  const previousDirPath = path.resolve(previousDir || '/');

  if (startDirPath === previousDirPath || fs.existsSync(tsConfigPath)) {
    return tsConfigPath;
  }

  return getPathToClosestTSConfig(path.join(startDir, '..'), startDir);
}

function getTSConfigPathFromConfig(globals: ConfigGlobals): string {
  const defaultTSConfigFile = getPathToClosestTSConfig();
  if (!globals) {
    return defaultTSConfigFile;
  }

  const tsJestConfig = getTSJestConfig(globals);

  if (tsJestConfig.tsConfigFile) {
    return tsJestConfig.tsConfigFile;
  }

  return defaultTSConfigFile;
}

export function mockGlobalTSConfigSchema(
  globals: ConfigGlobals,
): ConfigGlobals {
  const configPath = getTSConfigPathFromConfig(globals);
  return { 'ts-jest': { tsConfigFile: configPath } };
}

export const getTSConfig = _.memoize(getTSConfig_local, (globals, rootDir) => {
  // check cache before resolving configuration
  // NB: We use JSON.stringify() to create a consistent, unique signature. Although it lacks a uniform
  //     shape, this is simpler and faster than using the crypto package to generate a hash signature.
  return JSON.stringify(globals, rootDir);
});

// Non-memoized version of TSConfig
function getTSConfig_local(globals, rootDir: string = '') {
  const configPath = getTSConfigPathFromConfig(globals);
  logOnce(`Reading tsconfig file from path ${configPath}`);
  const skipBabel = getTSJestConfig(globals).skipBabel;

  const config = readCompilerOptions(configPath, rootDir);
  logOnce('Original typescript config before modifications: ', { ...config });

  // ts-jest will map lines numbers properly if inlineSourceMap and
  // inlineSources are set to true. The sourceMap configuration
  // is used to send the sourcemap back to Jest
  delete config.sourceMap;
  config.inlineSourceMap = true;
  config.inlineSources = true;

  // the coverage report is broken if `.outDir` is set
  // see https://github.com/kulshekhar/ts-jest/issues/201
  // `.outDir` is removed even for test files as it affects with breakpoints
  // see https://github.com/kulshekhar/ts-jest/issues/309
  delete config.outDir;

  if (configPath === path.join(getStartDir(), 'tsconfig.json')) {
    // hardcode module to 'commonjs' in case the config is being loaded
    // from the default tsconfig file. This is to ensure that coverage
    // works well. If there's a need to override, it can be done using
    // a custom tsconfig for testing
    config.module = tsc.ModuleKind.CommonJS;
  }

  config.module = config.module || tsc.ModuleKind.CommonJS;
  config.jsx = config.jsx || tsc.JsxEmit.React;

  if (config.allowSyntheticDefaultImports && !skipBabel) {
    // compile ts to es2015 and transform with babel afterwards
    config.module = tsc.ModuleKind.ES2015;
  }

  return config;
}

export function cacheFile(
  jestConfig: JestConfig,
  filePath: string,
  src: string,
): void {
  // store transpiled code contains source map into cache, except test cases
  if (!jestConfig.testRegex || !filePath.match(jestConfig.testRegex)) {
    const outputFilePath = path.join(
      jestConfig.cacheDirectory,
      '/ts-jest/',
      crypto
        .createHash('md5')
        .update(filePath)
        .digest('hex'),
    );

    fsExtra.outputFileSync(outputFilePath, src);
  }
}

export function runTsDiagnostics(
  filePath: string,
  compilerOptions: tsc.CompilerOptions,
): void {
  const program = tsc.createProgram([filePath], compilerOptions);
  const allDiagnostics = tsc.getPreEmitDiagnostics(program);

  if (allDiagnostics.length) {
    throw new Error(formatTsDiagnostics(allDiagnostics));
  }
}
