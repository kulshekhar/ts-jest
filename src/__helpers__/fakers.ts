import { Config } from '@jest/types'
import { resolve } from 'path'

import { createCompilerInstance } from '../compiler/instance'
import { ConfigSet } from '../config/config-set'
import { BabelConfig, TsCompiler, TsJestConfig, TsJestGlobalOptions } from '../types'
import { ImportReasons } from '../util/messages'

export function filePath(relPath: string): string {
  return resolve(__dirname, '..', '..', relPath)
}

export const rootDir = filePath('')

export function tsJestConfig(options?: Partial<TsJestConfig>): TsJestConfig {
  return {
    isolatedModules: false,
    compiler: 'typescript',
    transformers: [],
    babelConfig: undefined,
    tsConfig: undefined,
    packageJson: undefined,
    stringifyContentPathRegex: undefined,
    diagnostics: { ignoreCodes: [], pretty: false, throws: true },
    ...options,
  }
}

function getJestConfig<T extends Config.ProjectConfig>(
  options?: Partial<Config.InitialOptions | Config.ProjectConfig>,
  tsJestOptions?: TsJestGlobalOptions,
): T {
  const res = {
    globals: {},
    ...options,
  } as any
  if (tsJestOptions) {
    res.globals['ts-jest'] = tsJestOptions
  }

  return res
}

export function babelConfig<T extends BabelConfig>(options?: BabelConfig): T {
  return {
    ...options,
    presets: [...(options && options.presets)],
    plugins: [...(options && options.plugins)],
  } as any
}

export function importReason(text = '[[BECAUSE]]'): ImportReasons {
  return text as any
}

export const defaultResolve = (path: string): string => `resolved:${path}`

export function createConfigSet({
  jestConfig,
  tsJestConfig,
  parentConfig,
  resolve = defaultResolve,
  ...others
}: {
  jestConfig?: Partial<Config.ProjectConfig>
  tsJestConfig?: TsJestGlobalOptions
  parentConfig?: TsJestGlobalOptions
  resolve?: ((path: string) => string) | null
  [key: string]: any
} = {}): ConfigSet {
  const defaultTestRegex = ['(/__tests__/.*|(\\\\.|/)(test|spec))\\\\.[jt]sx?$']
  const defaultTestMatch = ['**/__tests__/**/*.[jt]s?(x)', '**/?(*.)+(spec|test).[jt]s?(x)']
  jestConfig = {
    ...jestConfig,
    testMatch: jestConfig?.testMatch ? [...jestConfig.testMatch, ...defaultTestMatch] : defaultTestMatch,
    testRegex: jestConfig?.testRegex ? [...defaultTestRegex, ...jestConfig.testRegex] : defaultTestRegex,
  }
  const cs = new ConfigSet(getJestConfig(jestConfig, tsJestConfig), parentConfig)
  if (resolve) {
    cs.resolvePath = resolve
  }
  Object.keys(others).forEach((key) => {
    Object.defineProperty(cs, key, { value: others[key] })
  })

  return cs
}

// not really unit-testing here, but it's hard to mock all those values :-D
export function makeCompiler({
  jestConfig,
  tsJestConfig,
  parentConfig,
}: {
  jestConfig?: Partial<Config.ProjectConfig>
  tsJestConfig?: TsJestGlobalOptions
  parentConfig?: TsJestGlobalOptions
} = {}): TsCompiler {
  tsJestConfig = { ...tsJestConfig }
  tsJestConfig.diagnostics = {
    ...(tsJestConfig.diagnostics as any),
    pretty: false,
  }
  const cs = createConfigSet({ jestConfig, tsJestConfig, parentConfig, resolve: null })

  return createCompilerInstance(cs)
}
