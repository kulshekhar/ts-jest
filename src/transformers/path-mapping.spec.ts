import { testing } from 'bs-logger'
import { join } from 'path'
import tsc from 'typescript'

import * as pathMapping from './path-mapping'
import { createConfigSet } from '../__helpers__/fakers'
import { normalizeSlashes } from '../utils/normalize-slashes'

const logger = testing.createLoggerMock()
const TS_JS_CODE_WITH_PATH_ALIAS = `
  import { parse } from '@utils/json'
  import hoo from '@utils/json'
  import * as json from '@utils/json'
  export * as jsonUtils from '@utils/json'
  const { stringify } = require('@utils/json')
  import foo = require('@utils/json')
  import('@utils/json').then(module => {
    module.parse('{foo:1}')
  })
  import type { Foo}  from '@utils/json'
  stringify({ foo: 1 })
  parse('{foo:1}')
  console.log(json)
  console.log(foo)
  const bar: Foo = { foo: 2 }
  hoo.foo(1)
`

describe('path-mapping', () => {
  test('should have correct signature', () => {
    expect(pathMapping.name).toBe('path-mapping')
    expect(typeof pathMapping.version).toBe('number')
    expect(pathMapping.version).toBeGreaterThan(0)
    expect(typeof pathMapping.factory).toBe('function')
  })

  test.each([
    {
      baseUrl: '.',
      paths: {
        '@utils/*': ['src/utils/*'],
      },
    },
    {
      rootDirs: ['./', 'foo'],
      baseUrl: '.',
      paths: {
        '@utils/*': ['src/utils/*'],
      },
    },
  ])(
    'should replace alias path with relative path which is resolved from paths tsconfig with js/ts extension',
    (tsConfig) => {
      const configSet = createConfigSet({
        tsJestConfig: {
          tsConfig,
        },
        logger,
      })
      const createFactory = () => pathMapping.factory(configSet)
      const transpile = (source: string) => tsc.transpileModule(source, { transformers: { before: [createFactory()] } })
      jest.spyOn(tsc, 'resolveModuleName').mockReturnValue({
        resolvedModule: {
          resolvedFileName: require.resolve('../utils/json'),
          extension: 'ts',
        } as any,
      })
      const out = transpile(TS_JS_CODE_WITH_PATH_ALIAS)

      expect(normalizeSlashes(out.outputText).replace(/\/\//g, '/')).toMatchSnapshot()

      jest.resetAllMocks()
    },
  )

  test.each([
    {
      code: `
        import styles from '@utils/app.css'

        console.log(styles)
      `,
      extension: 'css',
    },
    {
      code: `
        import jsonData from '@utils/data.json'

        console.log(jsonData)
      `,
      extension: 'json',
    },
    {
      code: `
        import vueComponent from '@utils/component.vue'

        console.log(vueComponent)
      `,
      extension: 'vue',
    },
  ])(
    'should replace alias path with relative path which is resolved from paths tsconfig with custom extensions',
    ({ code, extension }) => {
      const configSet = createConfigSet({
        tsJestConfig: {
          tsConfig: {
            baseUrl: '.',
            paths: {
              '@utils/*': ['src/utils/*'],
            },
          },
        },
        logger,
      })
      const createFactory = () => pathMapping.factory(configSet)
      const transpile = (source: string) => tsc.transpileModule(source, { transformers: { before: [createFactory()] } })
      const resolvedFileNameStub = join('..', `utils/json.${extension}`)
      jest.spyOn(tsc, 'resolveModuleName').mockReturnValue({
        resolvedModule: {
          resolvedFileName: resolvedFileNameStub,
          extension,
        } as any,
      })

      const out = transpile(code)

      expect(normalizeSlashes(out.outputText).replace(/\/\//g, '/')).toMatchSnapshot()

      jest.resetAllMocks()
    },
  )
})
