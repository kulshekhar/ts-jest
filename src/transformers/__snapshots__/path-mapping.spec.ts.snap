// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`path-mapping should replace alias path with relative path which is resolved from paths tsconfig with custom extensions 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
var app_css_1 = require(\\"../utils/json.css\\");
console.log(app_css_1.default);
"
`;

exports[`path-mapping should replace alias path with relative path which is resolved from paths tsconfig with custom extensions 2`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
var data_json_1 = require(\\"../utils/json.json\\");
console.log(data_json_1.default);
"
`;

exports[`path-mapping should replace alias path with relative path which is resolved from paths tsconfig with custom extensions 3`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
var component_vue_1 = require(\\"../utils/json.vue\\");
console.log(component_vue_1.default);
"
`;

exports[`path-mapping should replace alias path with relative path which is resolved from paths tsconfig with js/ts extension 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
exports.jsonUtils = void 0;
var json_1 = require(\\"./src/utils/json.ts\\");
var json = require(\\"./src/utils/json.ts\\");
exports.jsonUtils = require(\\"./src/utils/json.ts\\");
var stringify = require(\\"./src/utils/json.ts\\").stringify;
var foo = require(\\"./src/utils/json.ts\\");
Promise.resolve().then(function () { return require(\\"./src/utils/json.ts\\"); }).then(function (module) {
    module.parse('{foo:1}');
});
var json_2 = require(\\"./src/utils/json.ts\\");
stringify({ foo: 1 });
json_1.parse('{foo:1}');
console.log(json);
console.log(foo);
var bar = { foo: 2 };
"
`;

exports[`path-mapping should replace alias path with relative path which is resolved from paths tsconfig with js/ts extension 2`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
exports.jsonUtils = void 0;
var json_1 = require(\\"./src/utils/json.ts\\");
var json = require(\\"./src/utils/json.ts\\");
exports.jsonUtils = require(\\"./src/utils/json.ts\\");
var stringify = require(\\"./src/utils/json.ts\\").stringify;
var foo = require(\\"./src/utils/json.ts\\");
Promise.resolve().then(function () { return require(\\"./src/utils/json.ts\\"); }).then(function (module) {
    module.parse('{foo:1}');
});
var json_2 = require(\\"./src/utils/json.ts\\");
stringify({ foo: 1 });
json_1.parse('{foo:1}');
console.log(json);
console.log(foo);
var bar = { foo: 2 };
"
`;
